biorbd_optim/constraints.py

```diff
@@ -50,77 +50,112 @@ def add_constraints(ocp, nlp):
 
-            if elem[0] == Constraint.Type.MARKERS_TO_PAIR:
-                Constraint.__markers_to_pair(ocp, nlp, x, elem[2])
+    @staticmethod
+    def __get_instant(nlp, constraint):
+        if not isinstance(constraint["instant"], (list, tuple)):
+            constraint["instant"] = (constraint["instant"],)
+        x = MX()
+        u = MX()
+        last_node = False
+        for node in constraint["instant"]:
+            if isinstance(node, int):
+                if node < 0 or node > nlp["ns"]:
+                    raise RuntimeError("Invalid instant, " + str(node) + " must be between 0 and " + str(nlp["ns"]))
+                if node == nlp["ns"]:
+                    last_node = True
+                x = horzcat(x, nlp["X"][node])
+                u = horzcat(u, nlp["U"][node])
 
-            elif elem[0] == Constraint.Type.ALIGN_WITH_CUSTOM_RT:
-                Constraint.__align_with_custom_rt(ocp, nlp, x, elem[2])
+            elif node == Constraint.Instant.START:
+                x = horzcat(x, nlp["X"][0])
+                u = horzcat(u, nlp["U"][0])
 
-            elif elem[0] == Constraint.Type.PROPORTIONAL_Q:
-                Constraint.__proportional_variable(ocp, nlp, x, elem[2])
+            elif node == Constraint.Instant.MID:
+                if nlp["ns"] % 2 == 0:
+                    raise (ValueError("Number of shooting points must be odd to use MID"))
+                x = horzcat(x, nlp["X"][nlp["ns"] // 2 + 1])
+                u = horzcat(u, nlp["U"][nlp["ns"] // 2 + 1])
 
-            elif elem[0] == Constraint.Type.PROPORTIONAL_CONTROL:
-                if elem[1] == Constraint.Instant.END:
-                    raise RuntimeError("Instant.END is used even though there is no control u at last node")
-                Constraint.__proportional_variable(ocp, nlp, u, elem[2])
+            elif node == Constraint.Instant.INTERMEDIATES:
+                x = horzcat(x, nlp["X"][1 : nlp["ns"] - 1])
+                u = horzcat(u, nlp["U"][1 : nlp["ns"] - 1])
 
-            elif elem[0] == Constraint.Type.CONTACT_FORCE_GREATER_THAN:
-                if elem[1] == Constraint.Instant.END:
-                    raise RuntimeError("Instant.END is used even though there is no control u at last node")
-                Constraint.__contact_force_inequality(ocp, nlp, x, u, elem[2])
+            elif node == Constraint.Instant.END:
+                x = horzcat(x, nlp["X"][nlp["ns"]])
+
+            elif node == Constraint.Instant.ALL:
+                for i in range(nlp["ns"]):
+                    x = horzcat(x, nlp["X"][i])
+                    u = horzcat(u, nlp["U"][i])
+                x = horzcat(x, nlp["X"][nlp["ns"]])
+                last_node = True
+            else:
+                raise RuntimeError(" is not a valid instant")
+        return x, u, last_node
 
     @staticmethod
-    def __markers_to_pair(ocp, nlp, X, policy):
+    def __markers_to_pair(ocp, nlp, X, first_marker, second_marker):
         """
```

Comment:
You must assign a default value for them to work as dict I think

Reply:
Maybe I am wrong... If so I much prefer that since there is no default value to something where all values must be sent